File: src/CX/SesionUsuario.java
package CX;

public class SesionUsuario {
    private static SesionUsuario instancia;

    private Long usuarioId;
    private String nombreUsuario;
    private String rol;
    private String permisos;
    private boolean licenciaActiva;
    private String licenciaVencimiento;

    private SesionUsuario() {}

    public static SesionUsuario getInstancia() {
        if (instancia == null) {
            instancia = new SesionUsuario();
        }
        return instancia;
    }

    public void iniciarSesion(long id, String nombre, String rol, String permisos) {
        this.usuarioId = id;
        this.nombreUsuario = nombre;
        this.rol = rol;
        this.permisos = permisos;
    }

    public void iniciarSesion(long id, String nombre, String rol, String permisos,
                              boolean licenciaActiva, String licenciaVencimiento) {
        this.usuarioId = id;
        this.nombreUsuario = nombre;
        this.rol = rol;
        this.permisos = permisos;
        this.licenciaActiva = licenciaActiva;
        this.licenciaVencimiento = licenciaVencimiento;
    }

    public void cerrarSesion() {
        this.usuarioId = null;
        this.nombreUsuario = null;
        this.rol = null;
        this.permisos = null;
        this.licenciaActiva = false;
        this.licenciaVencimiento = null;
    }

    public Long getUsuarioId() {
        return usuarioId;
    }

    public String getNombreUsuario() {
        return nombreUsuario;
    }

    public String getRol() {
        return rol;
    }

    public String getPermisos() {
        return permisos;
    }

    public void setPermisos(String permisos) {
        this.permisos = permisos;
    }

    public boolean isLicenciaActiva() {
        return licenciaActiva;
    }

    public void setLicenciaActiva(boolean licenciaActiva) {
        this.licenciaActiva = licenciaActiva;
    }

    public String getLicenciaVencimiento() {
        return licenciaVencimiento;
    }

    public void setLicenciaVencimiento(String licenciaVencimiento) {
        this.licenciaVencimiento = licenciaVencimiento;
    }

    /** ¿El usuario actual tiene su licencia vencida (activa y con fecha pasada)? */
    public boolean licenciaVencida() {
        if (!licenciaActiva) return false;
        if (licenciaVencimiento == null || licenciaVencimiento.trim().isEmpty()) return false;
        try {
            java.time.LocalDate v = java.time.LocalDate.parse(licenciaVencimiento.trim(), java.time.format.DateTimeFormatter.ISO_LOCAL_DATE);
            return java.time.LocalDate.now().isAfter(v);
        } catch (Exception e) {
            return false;
        }
    }

    private String normalizar(String s) {
        if (s == null) return "";
        return s.toUpperCase()
                .replace("Á", "A").replace("É", "E").replace("Í", "I").replace("Ó", "O").replace("Ú", "U")
                .replaceAll("[^A-Z0-9:,]", ""); // Eliminar caracteres extraños o acentos no mapeados
    }

    public boolean tienePermiso(String modulo) {
        // Admin siempre tiene acceso total al modulo, ignorando permisos detallados.
        if ("Admin".equalsIgnoreCase(rol) || "Administrador".equalsIgnoreCase(rol)) return true;

        String p = normalizar(permisos);
        String m = normalizar(modulo);

        // El acceso al modulo (verlo en el menu) depende SOLO del permiso de "Ver" (V).
        // Desactivar "Ver" oculta el modulo aunque tenga permisos de accion (C/E/D/X).
        int idx = p.indexOf(m + ":");
        if (idx != -1) {
            int endIdx = p.indexOf(",", idx);
            String sub = (endIdx == -1) ? p.substring(idx) : p.substring(idx, endIdx);
            String ac = sub.substring(sub.indexOf(":") + 1);
            return ac.contains("V");
        }

        return p.contains(m);
    }

    public boolean tienePermiso(String modulo, String accion) {
        // Admin siempre tiene todos los permisos, ignorando permisos detallados.
        if ("Admin".equalsIgnoreCase(rol) || "Administrador".equalsIgnoreCase(rol)) return true;

        if (permisos == null) return false;
        
        String p = normalizar(permisos);
        String m = normalizar(modulo);
        String a = "";
        switch (accion.toUpperCase()) {
            case "VER": return tienePermiso(modulo); // Caso especial: Ver depende de tener cualquier permiso
            case "CREAR": a = "C"; break;
            case "EDITAR": a = "E"; break;
            case "ELIMINAR": a = "D"; break;
            case "EXPORTAR": a = "X"; break;
            default: a = accion.toUpperCase().substring(0, 1);
        }

        // Prioridad 1: Si existe una definición específica para el módulo (ej: PRODUCTOS:VCE)
        int idx = p.indexOf(m + ":");
        if (idx != -1) {
            int endIdx = p.indexOf(",", idx);
            String sub = (endIdx == -1) ? p.substring(idx) : p.substring(idx, endIdx);
            String actionsPart = sub.substring(sub.indexOf(":") + 1);
            return actionsPart.contains(a);
        }
        
        // Prioridad 2: Formato antiguo o permiso general por nombre
        return p.contains(m);
    }
}

File: inventario_visual/Core/SesionUsuario.vb
Namespace inventario_visual

    Public Module SesionUsuario

        Public Property UsuarioActual As Usuario

        Public Function TienePermiso(modulo As String) As Boolean
            If UsuarioActual Is Nothing Then Return False
            If EsAdministrador() Then Return True
            Dim permisos As String = Normalizar(UsuarioActual.Permisos)
            Dim moduloNormalizado As String = Normalizar(modulo)
            Dim inicio As Integer = permisos.IndexOf(moduloNormalizado & ":", StringComparison.Ordinal)
            If inicio >= 0 Then
                Dim fin As Integer = permisos.IndexOf(",", inicio, StringComparison.Ordinal)
                Dim acciones As String = If(fin < 0, permisos.Substring(inicio + moduloNormalizado.Length + 1), permisos.Substring(inicio + moduloNormalizado.Length + 1, fin - inicio - moduloNormalizado.Length - 1))
                Return acciones.Length > 0
            End If
            Return permisos.Contains(moduloNormalizado)
        End Function

        Public Function TienePermisoAccion(modulo As String, accion As String) As Boolean
            If UsuarioActual Is Nothing Then Return False
            If EsAdministrador() Then Return True
            Dim permisos As String = Normalizar(UsuarioActual.Permisos)
            Dim moduloNormalizado As String = Normalizar(modulo)
            Dim inicio As Integer = permisos.IndexOf(moduloNormalizado & ":", StringComparison.Ordinal)
            If inicio >= 0 Then
                Dim fin As Integer = permisos.IndexOf(",", inicio, StringComparison.Ordinal)
                Dim acciones As String = If(fin < 0, permisos.Substring(inicio + moduloNormalizado.Length + 1), permisos.Substring(inicio + moduloNormalizado.Length + 1, fin - inicio - moduloNormalizado.Length - 1))
                Dim codigo As String = If(accion.ToUpper() = "VER", "V", If(accion.ToUpper() = "CREAR", "C", If(accion.ToUpper() = "EDITAR", "E", If(accion.ToUpper() = "ELIMINAR", "D", If(accion.ToUpper() = "EXPORTAR", "X", accion.Substring(0, 1).ToUpper())))))
                Return acciones.Contains(codigo)
            End If
            Return permisos.Contains(String.Format("[{0}:{1}]", moduloNormalizado, accion.ToUpper())) OrElse permisos.Contains(moduloNormalizado)
        End Function

        Private Function EsAdministrador() As Boolean
            Return UsuarioActual IsNot Nothing AndAlso (String.Equals(UsuarioActual.Rol, "Admin", StringComparison.OrdinalIgnoreCase) OrElse String.Equals(UsuarioActual.Rol, "Administrador", StringComparison.OrdinalIgnoreCase))
        End Function

        Private Function Normalizar(valor As String) As String
            If String.IsNullOrEmpty(valor) Then Return ""
            Return valor.ToUpper().Replace("Á", "A").Replace("É", "E").Replace("Í", "I").Replace("Ó", "O").Replace("Ú", "U").Replace(" ", "")
        End Function

        Public Sub CerrarSesion()
            UsuarioActual = Nothing
        End Sub

    End Module

End Namespace

